R/rvar-.R

```diff
@@ -0,0 +1,1066 @@
+#' first dimension is used to index draws from the distribution.
+#' @template args-rvar-dim
+#' @template args-rvar-dimnames
+#' @template args-format-nchains
+#'
+#' @details
+#'
+#' The `"rvar"` class internally represents random variables as arrays of arbitrary
+#' dimension, where the first dimension is used to index draws from the distribution.
+#
+#' Most mathmetical operators and functions are supported, including efficient matrix
+#' multiplication and vector and array-style indexing. The intent is that an `rvar`
+#' works as closely as possible to how a base vector/matrix/array does, with a few
+#' differences:
+#'
+#' - The default behavior when subsetting is not to drop extra dimensions (i.e.
+#'   the default `drop` argument for `[` is `FALSE`, not `TRUE`).
+#' - Rather than base R-style recycling, `rvar`s use a limited form of broadcasting:
+#'   if an operation is being performed on two vectors with different size of the same
+#'   dimension, the smaller vector will be recycled up to the size of the larger one
+#'   along that dimension so long as it has size 1.
+#'
+#' For functions that expect base numeric arrays and for which `rvar`s cannot be
+#' used directly as arguments, you can use [rfun()] or [rdo()] to translate your
+#' code into code that executes across draws from one or more random variables
+#' and returns a random variable as output. Typically [rdo()] offers the most
+#' straightforward translation.
+#'
+#' As [rfun()] and [rdo()] incur some performance cost, you can also operate directly
+#' on the underlying array using the [draws_of()] function.
+#'
+#' @seealso [as_rvar()] to convert objects to `rvar`s.
+#'
+#' @return An object of class `"rvar"` representing a random variable.
+#'
+#' @export
+rvar <- function(x = double(), dim = NULL, dimnames = NULL, .nchains = 1L) {
+  x <- new_rvar(x, .nchains = .nchains)
+
+  if (!is.null(dim)) {
+    dim(x) <- dim
+  }
+  if (!is.null(dimnames)) {
+    dimnames(x) <- dimnames
+  }
+
+  x
+}
+
+#' @importFrom vctrs new_vctr
+new_rvar <- function(x = double(), .nchains = 1L) {
+  # TODO: decide on supported types and cast to them in here
+  .dim <- dim(x)
+  if (length(x) == 0) {
+    x <- double()
+  }
+  x <- as.array(x)
+
+  if (length(x) == 0) {
+    # canonical NULL rvar is 1 draw of nothing
+    # this ensures that (e.g.) extending a null rvar
+    # with x[1] = something works.
+    dim(x) <- c(1, 0)
+  }
+  else if (is.null(.dim) || length(.dim) == 1) {
+    # 1d vectors get treated as a single variable
+    dim(x) <- c(length(x), 1)
+  }
+
+  # ensure dimnames is set (makes comparison easier for tests)
+  if (length(dimnames(x)) == 0) {
+    dimnames(x) <- list(NULL)
+  }
+
+  .ndraws <- dim(x)[[1]]
+  .nchains <- as_one_integer(.nchains)
+  check_nchains_compat_with_ndraws(.nchains, .ndraws)
+
+  # ensure we have an index for draws
+  if (length(rownames(x)) == 0) {
+    rownames(x) <- as.character(seq_rows(x))
+  }
+
+  structure(
```

Comment:
I admit it feels a bit weird. Basically I experimented with several different formats, and this was the format that I was able to get to work with {vctrs} and the various {tidyverse} functions (e.g. so that you could include rvars in a tibble and functions like `tidyr::pivot_wider()` would still work). Other obvious candidates were not great for various reasons (a subclass of arrays didn't work properly with some tidyverse functions for reasons I couldn't quite figure out --- I think its length was measured incorrectly internally because some code bypassed the length() function I had overriden; an S4 class appeared to have other problems integrating with tibbles). The format I came down to still isn't perfect, and I think it in future it could be worth exploring different approach (maybe something with ALTREP?), but this format was the best I could do to get something working with the widest range of functions.